scopesim/source/source.py

```diff
@@ -505,93 +538,77 @@ def shift(self, dx=0, dy=0, layers=None):
-        spatial footprint. Source components instantiated from tables are shown as points.
+        Source components instantiated from 2d or 3d ImageHDUs are represented
+        by their spatial footprint. Source components instantiated from tables
+        are shown as points.
         """
         _, axes = figure_factory()
 
         colours = "rgbcymk" * (len(self.fields) // 7 + 1)
         for col, field in zip(colours, self.fields):
-            if isinstance(field, Table):
-                axes.plot(field["x"], field["y"], col+".")
-            elif isinstance(field, (fits.ImageHDU, fits.PrimaryHDU)):
-                xypts = imp_utils.calc_footprint(field.header)
-                convf = u.Unit(field.header["CUNIT1"]).to(u.arcsec)
-                outline = np.array(list(close_loop(xypts))) * convf
-                axes.plot(outline[:, 0], outline[:, 1], col)
-                axes.set_xlabel("x [arcsec]")
-                axes.set_ylabel("y [arcsec]")
+            field.plot(axes, col)
         axes.set_aspect("equal")
+        axes.set_xlabel("x [arcsec]")
+        axes.set_ylabel("y [arcsec]")
+        axes.legend()
+        return axes
 
     def make_copy(self):
         new_source = Source()
         new_source.meta = deepcopy(self.meta)
-        new_source._meta_dicts = deepcopy(self._meta_dicts)
-        new_source.spectra = deepcopy(self.spectra)
+        # new_source._meta_dicts = deepcopy(self._meta_dicts)
+        # new_source.spectra = deepcopy(self.spectra)
         for field in self.fields:
-            if isinstance(field, (fits.ImageHDU, fits.PrimaryHDU)) \
-                    and field._file is not None:  # and field._data_loaded is False:
-                new_source.fields.append(field)
-            else:
-                new_source.fields.append(deepcopy(field))
+            new_source.fields.append(deepcopy(field))
+            # TODO: The code below refers to DataContainer??
+            # if isinstance(field, (fits.ImageHDU, fits.PrimaryHDU)) \
+            #         and field._file is not None:  # and field._data_loaded is False:
+            #     new_source.fields.append(field)
+            # else:
+            #     new_source.fields.append(deepcopy(field))
 
         return new_source
 
     def append(self, source_to_add):
+        if not isinstance(source_to_add, Source):
+            raise ValueError(f"Cannot add {type(source_to_add)} object to Source object")
+
         new_source = source_to_add.make_copy()
         # If there is no field yet, then self._meta_dicts contains a
         # reference to self.meta, which is empty. This ensures that both are
         # updated at the same time. However, it is important that the fields
         # and _meta_dicts match when appending sources.
         if len(self.fields) == 0:
-            assert self._meta_dicts == [{}]
+            assert self._meta_dicts == [{}] or self._meta_dicts == []
```

Comment:
Can we fully remove the `_meta_dicts` attribute from the `Source` class now?

Reply:
That was the plan, yeah. Maybe keep it in along with the assert for now and see if that ever trips?

Reply:
done